optaplanner-core/src/main/java/org/optaplanner/core/impl/solver/DefaultSolverJob.java

```diff
@@ -108,7 +118,12 @@ public Solution_ call() {
             exceptionHandler.accept(problemId, e);
             throw new IllegalStateException("Solving failed for problemId (" + problemId + ").", e);
         } finally {
+            if (!solverStatusModifyingLock.isHeldByCurrentThread()) {
+                // reacquire the lock if we don't have it
+                solverStatusModifyingLock.lock();
+            }
             solvingTerminated();
+            solverStatusModifyingLock.unlock();
```

Comment:
This is not particularly safe, I'm afraid. What if `lock()` throws?  (I'd expect it can throw at least `InterruptedException`.)
WHat if  `solverTerminated()` throws?